Support for fuel gauge sensor MAX17048

maxpowel opened the pull request:
Hi, I am using this chip in my projects and I think that it is mature enought to be shared. And hopefully it can be useful for someone.

I also added a simple example. I tested it on NRF52 but it should work anywhere because it just requires a regular I2C.

maxpowel committed:
drivers: Added fuel gauge max17048

Added support for fuel gauge max17048

Signed-off-by: Alvaro Garcia <maxpowel@gmail.com>

teburd commented:
I'd note that Zephyr now has a fuel gauge API that this would better be represented by I believe. I don't think this is stale so removing that regardless.

teburd commented:
@aaronemassey please take a look and maybe make some suggestions on if this could be ported to the new fuel gauge api

aaronemassey commented:
> @aaronemassey please take a look and maybe make some suggestions on if this could be ported to the new fuel gauge api

It looks like all of the exposed sensor channels (one of the sensor channels doesn't exist anymore) already exist as a fuel gauge property. So I don't see any reason why this couldn't be ported to the fuel gauge API with relative ease.

I'd intended to do much of the driver migration myself (should be straight forward refactor) with reviews from their active maintainers.

aaronemassey commented on drivers/sensor/max17048/max17048.c:
```diff
@@ -0,0 +1,188 @@
+/* max17048.c - Driver for max17048 battery fuel gauge */
+
+/*
+ * Copyright (c) 2022 Alvaro Garcia Gomez <maxpowel@gmail.com>
```
nit: 2023 copyright :stuck_out_tongue:

aaronemassey commented on drivers/sensor/max17048/max17048.c:
```diff
@@ -0,0 +1,188 @@
+
+			val->val1 = whole_part;
+			val->val2 = decimal_part;
+		break;
```
The formatting for the break statements looks a little weird. Could you run a `clang-format`?

aaronemassey commented on drivers/sensor/max17048/max17048.c:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	uint16_t version = max17048_read_register(dev, REGISTER_VERSION) & 0xFFF0;
+	if (version != 0x10){
```
Seems like the version should be a device-tree property.

aaronemassey commented on drivers/sensor/max17048/max17048.c:
```diff
@@ -0,0 +1,188 @@
+	uint16_t version = max17048_read_register(dev, REGISTER_VERSION) & 0xFFF0;
+	if (version != 0x10){
+		printk("Wrong MAX17048 version, expected 0x10 but got %d. Maybe your wiring is wrong or it is a fake chip\n", version);
+		return -1;
```
Returning -1 might imply `-EPERM` error which doesn't quite make sense here. How about `-ENOTSUP`?


See `errno.h`

aaronemassey commented on drivers/sensor/max17048/max17048.h:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2022 Alvaro Garcia Gomez <maxpowel@gmail.com>
```
nit: Copyright here too.

aaronemassey commented on samples/sensor/max17048/Kconfig:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2022, Alvaro Garcia <maxpowel@gmail.com>
```
I don't think we need or want any of these KConfigs? None of the other drivers overwrite kconfigs.

maxpowel replied:
I agree

aaronemassey reviewed (changes requested):
Not a hard requirement, but I personally would prefer if you added emulated testing here. Based on the datasheet you can reuse a lot of the code from the `emul_sbs_gauge.c` since the `max17048` uses 16 byte words just like smbus.

aaronemassey commented on include/zephyr/drivers/sensor.h:
```diff
@@ -168,6 +168,8 @@ enum sensor_channel {
 	/** Time to full in minutes **/
 	SENSOR_CHAN_GAUGE_TIME_TO_FULL,
+	/** If the battery is being charged or discharged */
+	SENSOR_CHAN_GAUGE_CHARGING_STATE,
```
I don't think we want or need this. We can just check the charging current property.

maxpowel commented:
Thank you for your time. I would be glad to adapt it to the new API taking into account all your comments. 

I was reading these links about the new api
https://docs.zephyrproject.org/latest/hardware/peripherals/fuel_gauge.html
https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/include/zephyr/drivers/fuel_gauge.h
https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/drivers/fuel_gauge/sbs_gauge/sbs_gauge.h

But I'm not sure about how to implement this api. Is there any example that I can use for reference?

Thank you

aaronemassey commented:
> Thank you for your time. I would be glad to adapt it to the new API taking into account all your comments.
> 
> I was reading these links about the new api https://docs.zephyrproject.org/latest/hardware/peripherals/fuel_gauge.html https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/include/zephyr/drivers/fuel_gauge.h https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/drivers/fuel_gauge/sbs_gauge/sbs_gauge.h
> 
> But I'm not sure about how to implement this api. Is there any example that I can use for reference?
> 
> Thank you

Working prototype for the [SBS Compliant Fual Gauge driver](https://github.com/zephyrproject-rtos/zephyr/blob/513e03ad684810832968cfc8ed249b1898c6e92c/drivers/fuel_gauge/sbs_gauge/sbs_gauge.c)

Reminder: This API is still in the experimental stage, so some things are likely to change. I don't mind being responsible for those changes and migrations though.

maxpowel commented:
> Reminder: This API is still in the experimental stage, so some things are likely to change. I don't mind being responsible for those changes and migrations though.
No worries, I know that is is experimental and it is no problem if I have to do more changes in the future.

I have just adapted the code to the fuel gague api, updated the sample and added emul tests. I still want to improve a bit the code and the tests but can you tell me if the api is correctly implemented (in terms of structure)? 

I would be grateful if you can also take a look on the sample and tell me if this is how the fuel gauge api should be used.
https://github.com/zephyrproject-rtos/zephyr/blob/16043f62a40ab6c5f807d9dec5bc2f866bcf08c2/samples/fuel_gauge/max17048/src/main.c
Basically I call the function fuel_gauge_get_prop with a list of propts and then read the result.

Thanks

aaronemassey commented on drivers/fuel_gauge/max17048/emul_max17048.c:
```diff
@@ -0,0 +1,171 @@
+			if (msgs->len != 2) {
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+			}
+			uint16_t *value = (uint16_t *)msgs->buf;
```
Change

```
uint16_t *value = (uint16_t *)msgs->buf;
```
to
```
uint16_t value = sys_get_le16(msgs->buf);
```

This was a bug in the original emul_sbs_gauge emulator that I only found after using a big endian qemu.

See https://github.com/zephyrproject-rtos/zephyr/pull/55103

Thank you for contributing an emulator!!!

aaronemassey reviewed (changes requested):
This is great! Really appreciate the contribution of an additional fuel gauge driver and associated emulator.

aaronemassey commented:
> I would be grateful if you can also take a look on the sample and tell me if this is how the fuel gauge api should be used. https://github.com/zephyrproject-rtos/zephyr/blob/16043f62a40ab6c5f807d9dec5bc2f866bcf08c2/samples/fuel_gauge/max17048/src/main.c Basically I call the function fuel_gauge_get_prop with a list of propts and then read the result.
> 
> Thanks

Left some fairly minor comments. This is great, thank you!

maxpowel commented on tests/drivers/fuel_gauge/max17048/prj.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_I2C=y
+CONFIG_TEST_USERSPACE=y
+CONFIG_LOG=y
+CONFIG_SENSOR=y
```
I copied it from `https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/fuel_gauge/sbs_gauge/prj.conf` which has `CONFIG_SENSOR`.
It compiles perfectly without this so maybe in is not necessary there either.

Anyway, I removed it from this and after a few changes and everthing is ok

maxpowel commented:
I finished all the improvements I wanted to do. I also tested the example with a real device and it works perfectly!

aaronemassey commented on drivers/fuel_gauge/max17048/max17048.c:
```diff
@@ -0,0 +1,229 @@
+		return rc;
+	}
+
+	*response = (float)value * 0.208f;
```
Can we also document `0.208f`?

maxpowel replied:
Again, this weird values just comes from the datasheet
https://www.analog.com/media/en/technical-documentation/data-sheets/MAX17048-MAX17049.pdf
Page 11
Table 2. Register Summary (continued)
0.208%/hr

Sure I will document it.

aaronemassey commented on samples/fuel_gauge/max17048/README.rst:
```diff
@@ -0,0 +1,51 @@
+Overview
+********
+
+This sample shows how to use the Zephyr :ref:`sensor_api` API driver for the `MAX17048` sensor.
```
Should update this to `:ref:fuel_gauge_api`

aaronemassey commented on drivers/fuel_gauge/max17048/emul_max17048.c:
```diff
@@ -0,0 +1,170 @@
+	uint16_t addr;
+};
+
+static int emul_max17048_reg_write(const struct emul *target, int reg, int val)
```
Let's remove the commented out code and probably just return -EIO if we're not expecting a write

aaronemassey commented on drivers/fuel_gauge/max17048/emul_max17048.c:
```diff
@@ -0,0 +1,170 @@
+
+/** Run-time data used by the emulator */
+struct max17048_emul_data {
+	uint16_t mfr_acc;
```
We can delete the max17048_emul_data since it's unused.

aaronemassey commented on drivers/fuel_gauge/max17048/max17048.c:
```diff
@@ -0,0 +1,229 @@
+};
+
+/* Read buffer */
+unsigned char max17048_buffer[2];
```
Since we only use this in the `max17048_read_register()` function, let's make it local to that function.

aaronemassey commented on samples/fuel_gauge/max17048/README.rst:
```diff
@@ -0,0 +1,51 @@
+
+Features
+********
+By using this sensor you can get the following information:
```
Fuel guage is technically a sensor but might be a good idea to just change `sensor` to `fuel gauge` here in case someone confuses it with the `sensor` API.

aaronemassey commented on samples/fuel_gauge/max17048/README.rst:
````diff
@@ -0,0 +1,51 @@
+
+```
+*** Booting Zephyr OS build 16043f62a40a ***
+Found device "max17048@36", getting sensor data
````
`getting sensor` to `getting fuel gauge`, again just to avoid confusion with the sensor api.

aaronemassey reviewed (changes requested):
Really close! Just a few minor things. Also feel free to ping `aaronmassey#4455` on the Zephyr Discord if you have any questions and want a usually quicker response.

One thing I noticed was the usage of `floats` and many MCUs can't work with floats. But IMO that's not a blocker for the PR because it can always be changed in the future since the API works with ints.

maxpowel commented on drivers/fuel_gauge/max17048/max17048.c:
```diff
@@ -44,37 +43,59 @@ unsigned char max17048_buffer[2];
+	if (rc < 0){
+		return rc;
+	}
+	*response = (float)data * 78.125 / 1000000;
```
This thing comes from de datasheet
https://www.analog.com/media/en/technical-documentation/data-sheets/MAX17048-MAX17049.pdf
Page 10, Table 2. Register Summary:
78.125µV/cell
This ugly float operation is required when converting the value from the register into a voltage integer in the api shape.

Using integers would reduce precision

I will add a reference there as a comment for sure. I will review the code for more things like this

maxpowel commented on drivers/fuel_gauge/max17048/max17048.c:
```diff
@@ -44,37 +43,59 @@ unsigned char max17048_buffer[2];
+	}
+	*response = (float) data;
+
+	*response /= 256;
```
This thing comes from de datasheet in the same place as above
https://www.analog.com/media/en/technical-documentation/data-sheets/MAX17048-MAX17049.pdf
Page 10, Table 2. Register Summary:
1%/256 

I totally agree with you, people should not need to check the datasheet to understand how does it work. I will add it as a comment too

stephanosio commented:
> In any case, I did not want to reinvent the wheel implementing a software float processing because there are probably other easier options since this problem should be common

For most architectures, software floating point support is already implemented by the toolchain and there is no need for you to manually provide one.

maxpowel commented:
Thanks for the info @stephanosio

Shame on me, but just doing bit shifting I can avoid all the floats. The values sent by the IC are just integer values and only the constants are float

maxpowel commented:
I applied all the changes. Basically removed all float operations, cleaned coded based on your suggestions, added more documentation and updated the sensor references in the doc.

Everything compiles and run as expected (also tested in my devices)

aaronemassey commented on drivers/fuel_gauge/max17048/max17048.c:
```diff
@@ -62,39 +67,57 @@ int max17048_adc(const struct device *i2c_dev, uint16_t *response) {
+	Page 11, Table 2. Register Summary (continued): 0.208%/hr
+	
+	To avoid floats, the value will be multiplied by 208 instead of 0.208, taking into account
+	that the value will ve 1000 times higher
```
nit: `ve` to `be` :stuck_out_tongue:

maxpowel replied:
Oh, I checked it twice. Looks that I'm blind

aaronemassey commented on drivers/fuel_gauge/max17048/max17048.c:
```diff
@@ -39,17 +39,14 @@ struct max17048_data {
 int max17048_read_register(const struct device *dev, uint8_t registerId, uint16_t *response)
 {
-
+	unsigned char max17048_buffer[2];
```
Bah, I should've caught this earlier sorry. Can we make this a `uint8_t` array? That seems to be the style I've seen in Zephyr.

aaronemassey replied:
It's effectively the same, so I don't think there's any reason to retest.

aaronemassey reviewed (changes requested):
Awesome! One minor item that I missed last time.

After a bit of rebasing to fix commit history, I think we can get this merged pretty soon assuming @teburd approves.

aaronemassey commented:
Thanks for your patience with my reviews, I'm still getting used to GitHub's review format from Gerrit.

maxpowel commented:
> Thanks for your patience with my reviews, I'm still getting used to GitHub's review format from Gerrit.

No worries, take your time. There is not reason to rush. Thank you for your reviews.

I have just fixed your last rewiews.

aaronemassey reviewed (approved):
Lgtm, going to need to rebase the fixup commits and get @teburd 's view.

Thank you for the contribution and work!

I'll probably have to lgtm again after you rebase too.

maxpowel commented:
Thank you for your time, please tell me if you need something else from me.

By the way, just as a reminder, FUEL_GAUGE_INIT will need to be added once it is ready.

aaronemassey commented:
> By the way, just as a reminder, FUEL_GAUGE_INIT will need to be added once it is ready.

https://github.com/zephyrproject-rtos/zephyr/pull/55887 will probably merge fairly quick.

> please tell me if you need something else from me.

Main things I see blocking a merge are:

* One more +1 from another contributor, probably teburd.
* Rebase commits in branch to remove squash fixup commits for a cleaner history
* Rebase commits to remove all merge commits (Zephyr does rebasing over merging for a linear commit history)

aaronemassey reviewed (changes requested):
See https://github.com/zephyrproject-rtos/zephyr/pull/52349#issuecomment-1472319417

maxpowel commented:
I did the rebase stuff and also added the CONFIG_FUEL_GAUGE_INIT
Now I am dealing with all the style faults to pass all the workflows

maxpowel commented:
Ok now my code passes the compliance checks although there is an error related to "ZEPHYR_LIBLC3_MODULE" but I thinkg that this not related to me.

teburd reviewed (approved):
Looks good to me

If there's a sample per fuel gauge it'll get unwieldy pretty quickly as sensors saw, but I don't have a great recommendation other than perhaps as more come in create a common sample that uses overlays or maybe the future snippets feature @mbolivar is working on I think.

Board overlays seem like they might be hard here as most boards don't have a fuel gauge, its something you want to test out with like a breakout board probably.
